biodatasets/pubtator_central/pubtator_central.py

```diff
@@ -0,0 +1,291 @@
+                    "mentions": [
+                        {
+                            "concept_id": mention.id,
+                            "type": mention.type,
+                            "text": mention.text,
+                            "offsets": [mention.start, mention.end],
+                        }
+                        for mention in doc.annotations
+                    ],
+                }
+                yield source_example
+
+    def _pubtator_to_bigbio_kb(self, filepath: Dict) -> Iterator[Dict]:
+        with open(filepath, "r") as f:
+            unified_example = {}
+            for doc in pubtator.iterparse(f):
+                unified_example["id"] = doc.pmid
+                unified_example["document_id"] = doc.pmid
+
+                unified_example["passages"] = [
+                    {
+                        "id": doc.pmid + "_title",
+                        "type": "title",
+                        "text": [doc.title],
+                        "offsets": [[0, len(doc.title)]],
+                    },
+                    {
+                        "id": doc.pmid + "_abstract",
+                        "type": "abstract",
+                        "text": [doc.abstract],
+                        "offsets": [
+                            [
+                                # +1 assumes the title and abstract will be joined by a space.
+                                len(doc.title) + 1,
+                                len(doc.title) + 1 + len(doc.abstract),
+                            ]
+                        ],
+                    },
+                ]
+
+                unified_entities = {}
+                for entity in doc.annotations:
+                    # We need a unique identifier for this entity, so build it from the document id and entity id
+                    unified_entity_id = doc.pmid + "_" + entity.id
+                    # The user can provide a callable that returns the database name.
+                    db_name = self._get_db_name(entity)
+                    if unified_entity_id not in unified_entities:
+                        unified_entities[unified_entity_id] = {
+                            "id": unified_entity_id,
+                            "type": entity.type,
+                            "text": [entity.text],
+                            "offsets": [[entity.start, entity.end]],
+                            "normalized": [{"db_name": db_name, "db_id": entity.id}],
+                        }
+                    else:
+                        unified_entities[unified_entity_id]["text"].append(entity.text)
+                        unified_entities[unified_entity_id]["offsets"].append([entity.start, entity.end])
+                        unified_entities[unified_entity_id]["normalized"].append(
+                            {"db_name": db_name, "db_id": entity.id}
+                        )
+
+                unified_example["entities"] = list(unified_entities.values())
```

Comment:
I think our (confusing) way of handling offsets is causing some problems here ... examining the bigbio_kb output I see things like, 
```  {'id': '35157371_253959',
   'type': 'Gene',
   'text': ['TULIP-1', 'TULIP-1'],
   'offsets': [[3626, 3633], [4193, 4200]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'},
    {'db_name': 'ncbi_gene', 'db_id': '253959'}]},
```
the above should be two entity entries though, 
```
 {
   'id': '35157371_253959_a',
   'type': 'Gene',
   'text': ['TULIP-1'],
   'offsets': [[3626, 3633]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'}]
}, 
{
   'id': '35157371_253959_b',
   'type': 'Gene',
   'text': ['TULIP-1'],
   'offsets': [[4193, 4200]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'}]
},
```

(same for all entries in `entities`) we have text as a list and offsets as nested list to handle the rare cases of single entity mentions that are represented by discontinuous spans ... but each entity mention (in the case of pubtator) will point to a single span ... we just have to represent the text as a list and the offsets as a nested list to be consistent.

Reply:
Ah, I think I see, I was assuming entities with the same ID should get a single entry under `entities` and `text` and `offsets` should contain all of their unique mentions. I can fix this

Reply:
yes please!

Reply:
@sg-wbi just FYI

Reply:
Fixed here, will have to fix in #20/#370 later!

Reply:
Addressed here: https://github.com/bigscience-workshop/biomedical/pull/444